File: solutions/04.enterprise-bot/Middleware/Telemetry/TelemetryLuisRecognizer.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.ApplicationInsights;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.AI.Luis;
using Newtonsoft.Json.Linq;

namespace EnterpriseBot
{
    /// <summary>
    /// TelemetryLuisRecognizer invokes the Luis Recognizer and logs some results into Application Insights.
    /// Logs the Top Intent, Sentiment (label/score), (Optionally) Original Text
    /// Along with Conversation and ActivityID.
    /// The Custom Event name this logs is MyLuisConstants.IntentPrefix + "." + 'found intent name'
    /// For example, if intent name was "add_calender":
    ///    LuisIntent.add_calendar
    /// See <seealso cref="LuisRecognizer"/> for additional information.
    /// </summary>
    public class TelemetryLuisRecognizer : LuisRecognizer
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="TelemetryLuisRecognizer"/> class.
        /// </summary>
        /// <param name="application">The LUIS application to use to recognize text.</param>
        /// <param name="predictionOptions">The LUIS prediction options to use.</param>
        /// <param name="includeApiResults">TRUE to include raw LUIS API response.</param>
        /// <param name="logOriginalMessage">TRUE to include original user message.</param>
        /// <param name="logUserName">TRUE to include user name.</param>
        public TelemetryLuisRecognizer(LuisApplication application, LuisPredictionOptions predictionOptions = null, bool includeApiResults = false, bool logOriginalMessage = false, bool logUserName = false)
            : base(application, predictionOptions, includeApiResults)
        {
            LogOriginalMessage = logOriginalMessage;
            LogUsername = logUserName;
        }

        /// <summary>
        /// Gets a value indicating whether determines whether to log the Activity message text that came from the user.
        /// </summary>
        /// <value>If true, will log the Activity Message text into the AppInsight Custome Event for Luis intents.</value>
        public bool LogOriginalMessage { get; }

        /// <summary>
        /// Gets a value indicating whether determines whether to log the User name.
        /// </summary>
        /// <value>If true, will log the user name into the AppInsight Custom Event for Luis intents.</value>
        public bool LogUsername { get; }

        /// <summary>
        /// Analyze the current message text and return results of the analysis (Suggested actions and intents).
        /// </summary>
        /// <param name="context">Context object containing information for a single turn of conversation with a user.</param>
        /// <param name="ct">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
        /// <param name="logOriginalMessage">Determines if the original message is logged into Application Insights.  This is a privacy consideration.</param>
        /// <returns>The LUIS results of the analysis of the current message text in the current turn's context activity.</returns>
        public async Task<RecognizerResult> RecognizeAsync(ITurnContext context, CancellationToken ct, bool logOriginalMessage = false)
        {
            if (context == null)
            {
                throw new ArgumentNullException(nameof(context));
            }

            // Call Luis Recognizer
            var recognizerResult = await base.RecognizeAsync(context, ct);

            var conversationId = context.Activity.Conversation.Id;

            // Find the Telemetry Client
            if (context.TurnState.TryGetValue(TelemetryLoggerMiddleware.AppInsightsServiceKey, out var telemetryClient) && recognizerResult != null)
            {
                var topLuisIntent = recognizerResult.GetTopScoringIntent();
                var intentScore = topLuisIntent.score.ToString("N2");

                // Add the intent score and conversation id properties
                var telemetryProperties = new Dictionary<string, string>()
                {
                    { LuisTelemetryConstants.ActivityIdProperty, context.Activity.Id },
                    { LuisTelemetryConstants.IntentProperty, topLuisIntent.intent },
                    { LuisTelemetryConstants.IntentScoreProperty, intentScore },
                };

                if (recognizerResult.Properties.TryGetValue("sentiment", out var sentiment) && sentiment is JObject)
                {
                    if (((JObject)sentiment).TryGetValue("label", out var label))
                    {
                        telemetryProperties.Add(LuisTelemetryConstants.SentimentLabelProperty, label.Value<string>());
                    }

                    if (((JObject)sentiment).TryGetValue("score", out var score))
                    {
                        telemetryProperties.Add(LuisTelemetryConstants.SentimentScoreProperty, score.Value<string>());
                    }
                }

                if (!string.IsNullOrEmpty(conversationId))
                {
                    telemetryProperties.Add(LuisTelemetryConstants.ConversationIdProperty, conversationId);
                }

                // Add Luis Entitites
                var entities = new List<string>();
                foreach (var entity in recognizerResult.Entities)
                {
                    if (!entity.Key.ToString().Equals("$instance"))
                    {
                        entities.Add($"{entity.Key}: {entity.Value.First}");
                    }
                }

                // For some customers, logging user name within Application Insights might be an issue so have provided a config setting to disable this feature
                if (logOriginalMessage && !string.IsNullOrEmpty(context.Activity.Text))
                {
                    telemetryProperties.Add(LuisTelemetryConstants.QuestionProperty, context.Activity.Text);
                }

                // Track the event
                ((TelemetryClient)telemetryClient).TrackEvent($"{LuisTelemetryConstants.IntentPrefix}.{topLuisIntent.intent}", telemetryProperties);
            }

            return recognizerResult;
        }
    }
}

File: samples/javascript_typescript/52.enterprise-bot/src/middleware/telemetry/telemetryLuisRecognizer.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License

import { TelemetryClient } from "applicationinsights";
import { RecognizerResult, TurnContext } from "botbuilder";
import { LuisApplication, LuisPredictionOptions, LuisRecognizer } from "botbuilder-ai";
import { LuisTelemetryConstants } from "./luisTelemetryConstants";
import { TelemetryLoggerMiddleware } from "./telemetryLoggerMiddleware";

/**
 * TelemetryLuisRecognizer invokes the Luis Recognizer and logs some results into Application Insights.
 * Logs the Top Intent, Sentiment (label/score), (Optionally) Original Text Along with Conversation
 * and ActivityID.
 * The Custom Event name this logs is MyLuisConstants.IntentPrefix + "." + 'found intent name'
 * For example, if intent name was "add_calender": LuisIntent.add_calendar
 */
export class TelemetryLuisRecognizer extends LuisRecognizer {
    private readonly _logOriginalMessage: boolean;
    private readonly _logUsername: boolean;

    /**
     * Initializes a new instance of the TelemetryLuisRecognizer class.
     * @param {LuisApplication} application The LUIS application to use to recognize text.
     * @param {LuisPredictionOptions} predictionOptions The LUIS prediction options to use.
     * @param {boolean} includeApiResults TRUE to include raw LUIS API response.
     * @param {boolean} logOriginalMessage TRUE to include original user message.
     * @param {boolean} logUserName TRUE to include user name.
     */
    constructor(application: LuisApplication, predictionOptions?: LuisPredictionOptions, includeApiResults: boolean = false, logOriginalMessage: boolean = false, logUserName: boolean = false) {
        super(application, predictionOptions, includeApiResults);
        this._logOriginalMessage = logOriginalMessage;
        this._logUsername = logUserName;
    }

    /**
     * Gets a value indicating whether determines whether to log the Activity message text that came from the user.
     */
    public get logOriginalMessage(): boolean { return this._logOriginalMessage; }

    /**
     * Gets a value indicating whether determines whether to log the User name.
     */
    public get logUsername(): boolean { return this._logUsername; }

    /**
     * Analyze the current message text and return results of the analysis (Suggested actions and intents).
     * @param {TurnContext} context Context object containing information for a single turn of conversation with a user.
     * @param {boolean} logOriginalMessage Determines if the original message is logged into Application Insights. This is a privacy consideration.
     */
    public async recognize(context: TurnContext, logOriginalMessage: boolean = false): Promise<RecognizerResult> {
        if (context === null) {
            throw new Error("context is null");
        }

        // Call Luis Recognizer
        const recognizerResult: RecognizerResult = await super.recognize(context);

        const conversationId: string = context.activity.conversation.id;

        // Find the Telemetry Client
        if (recognizerResult && context.turnState.has(TelemetryLoggerMiddleware.AppInsightsServiceKey)) {
            const telemetryClient: TelemetryClient = context.turnState.get(TelemetryLoggerMiddleware.AppInsightsServiceKey);

            const topLuisIntent: string = LuisRecognizer.topIntent(recognizerResult);
            const intentScore: number = recognizerResult.intents[topLuisIntent].score;

            // Add the intent score and conversation id properties
            const properties: { [key: string]: string } = {};
            properties[LuisTelemetryConstants.ActivityIdProperty] = context.activity.id || "";
            properties[LuisTelemetryConstants.IntentProperty] = topLuisIntent;
            properties[LuisTelemetryConstants.IntentScoreProperty] = intentScore.toString();

            if (recognizerResult.sentiment) {
                if (recognizerResult.sentiment.label) {
                    properties[LuisTelemetryConstants.SentimentLabelProperty] = recognizerResult.sentiment.label;
                }

                if (recognizerResult.sentiment.score) {
                    properties[LuisTelemetryConstants.SentimentScoreProperty] = recognizerResult.sentiment.score.toString();
                }
            }

            if (conversationId) {
                properties[LuisTelemetryConstants.ConversationIdProperty] = conversationId;
            }

            // For some customers, logging user name within Application Insights might be an issue so have provided a config setting to disable this feature
            if (logOriginalMessage && context.activity.text) {
                properties[LuisTelemetryConstants.QuestionProperty] = context.activity.text;
            }

            // Track the event
            telemetryClient.trackEvent({
                name: `${LuisTelemetryConstants.IntentPrefix}.${topLuisIntent}`,
                properties,
            });
        }

        return recognizerResult;
    }
}
